apis/v1alpha1/tlsroute_types.go

```diff
@@ -0,0 +1,131 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// TLSRouteSpec defines the desired state of TLSRoute
+type TLSRouteSpec struct {
+	// Rules are a list of TLS matchers and actions.
+	// +required
+	// +kubebuilder:validation:MinItems=1
+	Rules []TLSRouteRule `json:"rules" protobuf:"bytes,1,rep,name=rules"`
+}
+
+// TLSRouteStatus defines the observed state of TLSRoute
+type TLSRouteStatus struct {
+	GatewayRefs []GatewayObjectReference `json:"gatewayRefs" protobuf:"bytes,1,rep,name=gatewayRefs"`
+}
+
+// +genclient
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+
+// TLSRoute is the Schema for the TlsRoutes API.
+type TLSRoute struct {
+	metav1.TypeMeta   `json:",inline" protobuf:"bytes,1,opt,name=typeMeta"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,2,opt,name=metadata"`
+
+	Spec   TLSRouteSpec   `json:"spec,omitempty" protobuf:"bytes,3,opt,name=spec"`
+	Status TLSRouteStatus `json:"status,omitempty" protobuf:"bytes,4,opt,name=status"`
+}
+
+// TLSRouteRule is the configuration for a given rule.
+type TLSRouteRule struct {
```

Comment:
Shall we have this (and TCPRoute too I guess) follow the HTTPRoute pattern by making the `Matched` field a slice and adding a `Filter` field?

Reply:
Making Match a slice is a good update. I'll update it here and send another PR to update it for TCPRoute.

Adding a filter seems pre-mature. I'd like to know use-cases and and an actual need for it before we add it. If we can do it after v1alpha1, that would be better.

Reply:
> send another PR to update it for TCPRoute.

So, TCPRoute has a match but it only has an extensionRef, which can have plural matches if needed. The counter argument is that we should have a uniformity in our APIs.

Reply:
IMHO we should adjust everything to have:
```Go
Matches []XXXRouteMatch `json:"matches" protobuf:"bytes,1,rep,name=matches"`
```

Reply:
I agree: https://github.com/kubernetes-sigs/service-apis/pull/282

Reply:
The 'Matches` part of this thread seems to be resolved now.

Regarding a `Filter` field, do we have any use cases that require `Filter`? I'd like to add more fields to the route incrementally (in a backwards compatible way) after this PR is merged in.

Reply:
Yeh if we don't have any obvious filters now, let's table that discussion.